### PR TITLE
Deploy script take 2

### PR DESCRIPTION
**Pull request by ChihChengLiang**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,7 +7,8 @@
     "test": "buidler test",
     "lint": "prettier --check \"**/*.{sol,ts,js}\"",
     "prettier": "prettier --write \"**/*.{sol,ts,js}\"",
-    "generate": "buidler compile && typechain --target truffle './build/**/*.json'",
+    "generate": "buidler compile && typechain --target truffle './build/**/*.json' && typechain --target ethers-v4 './build/**/*.json'",
```
Comment (ChihChengLiang) [inline package.json:10]: keep truffle typechain for backward compatibility

File: test/bls.test.ts [hunk 2]
```diff
@@ -1,22 +1,32 @@
...
 
 describe("BLS", async () => {
-    let bls: TestBlsInstance;
+    let bls: TestBls;
     before(async function() {
         await mcl.init();
-        bls = await TestBLS.new();
+        const accounts = await ethers.getSigners();
+        bls = await new TestBlsFactory(accounts[0]).deploy();
+        await bls.deployed();
```
Comment (ChihChengLiang) [inline test/bls.test.ts:15]: This is an example of using buidler/typechain ether-v4. Note that:
- We get signers from the buidler environment. We `import { ethers } from "@nomiclabs/buidler"`
- We use the factory generated by typechain/ethers-v4 to deploy contract.
- Remember to `await contact.deployed();`

File: test/bls.test.ts [hunk 3]
```diff
@@ -1,22 +1,32 @@
...
     });
     it("hash to point", async function() {
         for (let i = 0; i < 20; i++) {
             const data = web3.utils.randomHex(12);
             let expect = mcl.g1ToHex(mcl.hashToPoint(data));
-            let res = await bls.hashToPoint(data);
-            assert.equal(expect[0], bnToHex(res[0]));
-            assert.equal(expect[1], bnToHex(res[1]));
+            let res = await bls.hashToPoint(ethers.utils.arrayify(data));
```
Comment (ChihChengLiang) [inline test/bls.test.ts:21]: The contract functions now take bigNumberish for integer types. 
But for the "bytes" type, we need to turn the hex into `arrayish` by using `ethers.utils.arrayify`

File: test/bls.test.ts [hunk 4]
```diff
@@ -294,7 +251,7 @@ describe("BLS", async () => {
         let messages_ser = messages.map(p => mcl.g1ToBN(p));
         let pubkeys_ser = pubkeys.map(p => mcl.g2ToBN(p));
         let sig_ser = mcl.g1ToBN(aggSignature);
-        let cost = await bls.verifyMultipleGasCost.call(
+        let cost = await bls.estimateGas.verifyMultipleGasCost(
```
Comment (ChihChengLiang) [inline test/bls.test.ts:254]: This is how we write estimateGas call

File: ts/deploy.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+import { IncrementalTreeFactory } from "../types/ethers-contracts/IncrementalTreeFactory";
+import { RollupFactory } from "../types/ethers-contracts/RollupFactory";
+import { RollupRedditFactory } from "../types/ethers-contracts/RollupRedditFactory";
+
+import { ethers, Signer, Contract } from "ethers";
+
+interface DeploymentParameters {
+    MAX_DEPTH: number;
+    MAX_DEPOSIT_SUBTREE_DEPTH: number;
+}
```
Comment (ChihChengLiang) [inline ts/deploy.ts:28]: We manage the deployment parameters with an interface like this.

File: ts/deploy.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    const logger = await new LoggerFactory(signer).deploy();
+    await logger.deployed();
+    await nameRegistry.registerName(
+        await paramManager.LOGGER(),
+        logger.address
+    );
+
+    const allLinkRefs = {
+        __$b941c30c0f5422d8b714f571f17d94a5fd$__: paramManager.address,
+        __$a6b8846b3184b62d6aec39d1f36e30dab3$__: rollupUtils.address
```
Comment (ChihChengLiang) [inline ts/deploy.ts:69]: linking is annoying. But the factory from the typechain/ethers-v4 requires this exact inputs of linking reference.

These are produced by `solidityHash(filepath+libraryname)`

File: package.json [hunk 7]
```diff
@@ -7,7 +7,8 @@
...
+    "deploy": "ts-node ./ts/deploy.ts",
```
Comment (ChihChengLiang) [inline package.json:11]: Use this script to deploy contracts.

File: test/bls.test.ts [hunk 8]
```diff
@@ -32,19 +42,19 @@ describe("BLS", async () => {
             messages.push(M);
             pubkeys.push(pubkey);
         }
-        let messages_ser = messages.map(p => mcl.g1ToBN(p));
-        let pubkeys_ser = pubkeys.map(p => mcl.g2ToBN(p));
-        let sig_ser = mcl.g1ToBN(aggSignature);
+        let messages_ser = messages.map(p => mcl.g1ToHex(p));
+        let pubkeys_ser = pubkeys.map(p => mcl.g2ToHex(p));
+        let sig_ser = mcl.g1ToHex(aggSignature);
```
Comment (ChihChengLiang) [inline test/bls.test.ts:47]: Ethersjs wraps bn.js with BigNumber, but the latter is incompatible with the former. So we can't feed bn.js to ethers contract functions.
However, ethers contract functions accept hex as bigNumberish.

File: ts/deploy.ts (hunk 6 above)
Comment (vaibhavchellani) [reply ts/deploy.ts:69]: Is there no workaround for this? :((.

File: tsconfig.json [hunk 9]
```diff
@@ -18,6 +18,9 @@
   "include": ["**/*.ts"],
   "exclude": ["node_modules", "example", "build"],
   "files": [
-    "./buidler.config.ts"
+    "./buidler.config.ts",
+    "./node_modules/@nomiclabs/buidler-ethers/src/type-extensions.d.ts",
```
Comment (vaibhavchellani) [inline tsconfig.json:22]: Do we want to use ethers and web3 both?

**Pull request merged**